### PR TITLE
Fixes for concurrency control

### LINKED ISSUE #61300: Setting `concurrent_threads_soft_limit_num` doesn't work
**Describe what's wrong**
It is expected that the `concurrent_threads_soft_limit_num` setting will limit the maximum number of threads used to process a request,  according to the docs [one](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#concurrent_threads_soft_limit_num) [two](https://clickhouse.com/docs/en/development/architecture#concurrency-control)

It is observed that no restriction occurs.

**Does it reproduce on the most recent release?**
Yes, checked on version 24.2.1.2248

**Enable crash reporting**

**How to reproduce**

> Which ClickHouse server version to use

 Reproduced in 24.2.1.2248
> Non-default settings, if any

Set `<concurrent_threads_soft_limit_num>2</concurrent_threads_soft_limit_num>` in config.xml
Set ` <max_threads>32</max_threads>` in users.xml
> Queries to run that lead to an unexpected result

Run query `SELECT count(*) FROM numbers_mt(100000000);` and remember its query_id.

Get thread information for this query: 
```sql
select
   length(thread_ids) as total_threads,
   peak_threads_usage
from system.query_log 
where 
   type = 'QueryFinish' 
   and query_id = '{previous_querry_id}';
```

Output:
```
┌─total_threads─┬─peak_threads_usage─┐
│            34 │                 34 │
└───────────────┴────────────────────┘
```

**Expected behavior**

The number of threads is expected to be around 2 (+2) as set in `concurrent_threads_soft_limit_num`, but we see that it is about `max_threads`.

**Error message and/or stacktrace**

None

**Additional context**

Found that the integration test for these setting was disabled as broken https://github.com/ClickHouse/ClickHouse/commit/c7e058fa6e017a7539e702f87dd7b94d8c53c797.

### PR DESCRIPTION
**Pull request by serxa**
The problem is due to improper `use_concurrency_control` setting handling. All queries do their CPU slot allocations with `min_threads` = `max_threads`, basically ignoring CPU slot limit.

If concurrency control is not used we should not even count threads as competing. To avoid counting them in ConcurrencyControl, we now create a dummy slot allocation. This way, for example, a merge thread cannot lower the number of CPU slots available for regular queries.

Fixes https://github.com/ClickHouse/ClickHouse/issues/61300

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `use_concurrency_control` setting handling for proper `concurrent_threads_soft_limit_num` limit enforcing. This enables concurrency control by default because previously it was broken.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit dd6c1ac19addfe80de6dfc6b380b93fb963632f9 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61473/dd6c1ac19addfe80de6dfc6b380b93fb963632f9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by Arnoways** [comment]
Hi, 
Any chance this bug fix might get some love ? 
It would be awesome to get this setting to work.

**Comment by serxa** [comment]
Yes, I'm going to return to fixing it this week

**Comment by serxa** [comment]
It should be fixed now. Let's check if CI objects

**Comment by serxa** [comment]
`02845_threads_count_in_distributed_queries` and `01091_num_threads` failed. Now that we fixed the feature it started to work. I need to disable it in all the tests that might be affected and rely on the number of threads.

**Comment by serxa** [comment]
Enabled tet fails now with:
```
>       assert check(
            int(
                node.query(
                    f"select ProfileEvents['{profile_event}'] from system.query_log where current_database = currentDatabase() and query_id = '{query_id}' and type = 'QueryFinish' order by query_start_time_microseconds desc limit 1"
                )
            )
        )
E       assert False
E        +  where False = <function test_concurrent_threads_soft_limit_defined_50.<locals>.<lambda> at 0x7f3254144b80>(68)
E        +    where 68 = int('68\n')
E        +      where '68\n' = <bound method ClickHouseInstance.query of <helpers.cluster.ClickHouseInstance object at 0x7f32542f6c50>>("select ProfileEvents['ConcurrencyControlGrantDelayed'] from system.query_log where current_database = currentDatabase() and query_id = 'test_concurrent_threads_soft_limit_2' and type = 'QueryFinish' order by query_start_time_microseconds desc limit 1")
E        +        where <bound method ClickHouseInstance.query of <helpers.cluster.ClickHouseInstance object at 0x7f32542f6c50>> = <helpers.cluster.ClickHouseInstance object at 0x7f32542f6c50>.query

```

**Comment by serxa** [comment]
It feels that 68 slots (instead of only 50) were busy at the moment of query start. Which means that 18 slots were most probably acquire for some other activity (not a query), which is wrong. Let me add logs to identify that activity.

**Comment by serxa** [comment]
Now the problem is different:
```
2024.08.03 00:40:03.549435 [ 1 ] {} <Information> Application: ConcurrencyControl limit is set to 16
```
For sume reason limit for number of CPUs available in the system wins over the limit set by `<concurrent_threads_soft_limit_num>50</concurrent_threads_soft_limit_num>` server setting. Fixing...

**Comment by serxa** [comment]
Problems with CI:
```
2024.08.29 23:21:25.992734 [ 1155 ] {} <Fatal> : Logical error: 'Received from kng4alm55c.us-east-2.aws.clickhouse-staging.com:9440. DB::Exception: Unexpected error for async deduplicated insert, please check error logs. Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cb5a404
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007f004fc
2. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000007f162b4
3. DB::MergeTreeDataWriter::splitBlockIntoParts(DB::Block&&, unsigned long, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::Context const>, std::shared_ptr<DB::AsyncInsertInfo>) @ 0x00000000115ad838
4. DB::SharedMergeTreeSinkImpl<true>::consume(DB::Chunk&) @ 0x0000000011765870
5. DB::SinkToStorage::onConsume(DB::Chunk) @ 0x0000000011cf0778
6. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::ExceptionKeepingTransform::work()::$_1, void ()>>(std::__function::__policy_storage const*) @ 0x0000000011c4563c
7. DB::runStep(std::function<void ()>, DB::ThreadStatus*, std::atomic<unsigned long>*) @ 0x0000000011c452f8
8. DB::ExceptionKeepingTransform::work() @ 0x0000000011c44ae4
9. DB::ExecutionThreadContext::executeTask() @ 0x0000000011a31eb0
10. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000011a285f0
11. DB::PipelineExecutor::execute(unsigned long, bool) @ 0x0000000011a27cc8
12. DB::CompletedPipelineExecutor::execute() @ 0x0000000011a267ac
13. DB::AsynchronousInsertQueue::processData(DB::AsynchronousInsertQueue::InsertQuery, std::unique_ptr<DB::AsynchronousInsertQueue::InsertData, std::default_delete<DB::AsynchronousInsertQueue::InsertData>>, std::shared_ptr<DB::Context const>, DB::AsynchronousInsertQueue::QueueShardFlushTimeHistory&) @ 0x000000000fee8568
14. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::AsynchronousInsertQueue::scheduleDataProcessingJob(DB::AsynchronousInsertQueue::InsertQuery const&, std::unique_ptr<DB::AsynchronousInsertQueue::InsertData, std::default_delete<DB::AsynchronousInsertQueue::InsertData>>, std::shared_ptr<DB::Context const>, unsigned long)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000feef9c4
15. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000cc0359c
16. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000cc06ff0
17. void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000cc05e0c
18. ? @ 0x000000000007d5c8
19. ? @ 0x00000000000e5edc
'.
```

**Comment by serxa** [comment]
It is not clear what exactly failed in tidy build but it is related to sccache:
```
Sep 16 19:11:30 sccache: error: Server startup failed: cache storage failed to read: PermissionDenied (permanent) at stat => no valid credential found and anonymous access is not allowed
Sep 16 19:11:30 
Sep 16 19:11:30 Context:
Sep 16 19:11:30    service: s3
Sep 16 19:11:30    path: .sccache_check
Sep 16 19:11:30 
Sep 16 19:11:30 Backtrace:
Sep 16 19:11:30    0: <unknown>
Sep 16 19:11:30    1: <unknown>
Sep 16 19:11:30    2: <unknown>
Sep 16 19:11:30    3: <unknown>
Sep 16 19:11:30    4: <unknown>
Sep 16 19:11:30    5: <unknown>
Sep 16 19:11:30    6: <unknown>
Sep 16 19:11:30    7: <unknown>
Sep 16 19:11:30    8: <unknown>
Sep 16 19:11:30    9: <unknown>
Sep 16 19:11:30   10: <unknown>
Sep 16 19:11:30   11: <unknown>
Sep 16 19:11:30   12: <unknown>
Sep 16 19:11:30   13: <unknown>
Sep 16 19:11:30   14: <unknown>
Sep 16 19:11:30   15: <unknown>
Sep 16 19:11:30   16: <unknown>
Sep 16 19:11:30   17: <unknown>
Sep 16 19:11:30   18: <unknown>
Sep 16 19:11:30   19: <unknown>
Sep 16 19:11:30   20: <unknown>
Sep 16 19:11:30   21: <unknown>
Sep 16 19:11:30   22: <unknown>
Sep 16 19:11:30 
Sep 16 19:11:30 
Sep 16 19:11:30 Run with SCCACHE_LOG=debug SCCACHE_NO_DAEMON=1 to get more information
Sep 16 19:11:38 [14076/14105] Linking CXX executable src/Common/ZooKeeper/examples/zkutil_test_async
Sep 16 19:11:41 [14077/14105] Linking CXX executable src/Common/ZooKeeper/examples/zkutil_test_commands_new_lib
Sep 16 19:11:41 [14078/14105] Linking CXX executable src/Storages/examples/get_current_inserts_in_replicated
Sep 16 19:11:42 [14079/14105] Linking CXX executable programs/clickhouse-odbc-bridge
Sep 16 19:11:45 [14080/14105] Linking CXX executable programs/clickhouse-library-bridge
Sep 16 19:11:48 [14081/14105] Linking CXX executable utils/keeper-bench/keeper-bench
Sep 16 19:12:05 [14082/14105] Linking CXX executable src/unit_tests_dbms
Sep 16 19:12:05 ninja: build stopped: cannot make progress due to previous errors.
Traceback (most recent call last):
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/docker/packager/./packager", line 559, in <module>
    main()
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/docker/packager/./packager", line 546, in main
    run_docker_image_with_env(
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/docker/packager/./packager", line 115, in run_docker_image_with_env
    subprocess.check_call(cmd, shell=True)
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'docker run --network=host --user=1000:1000 --rm  --volume=/home/ubuntu/actions-runner/_work/_temp/build_check/binary_tidy:/output --volume=/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse:/build  -e OUTPUT_DIR=/output -e DEB_ARCH=amd64 -e CC=clang-18 -e CXX=clang++-18 -e BUILD_TYPE=Debug -e SCCACHE_BUCKET=clickhouse-builds -e SCCACHE_S3_KEY_PREFIX=ccache/sccache -e CTCACHE_S3_BUCKET=clickhouse-builds -e CTCACHE_S3_FOLDER=ccache/clang-tidy-cache -e NINJA_FLAGS=-k0 -e VERSION_STRING='24.9.1.2758' -e CMAKE_FLAGS="$CMAKE_FLAGS -DCMAKE_C_COMPILER=clang-18 -DCMAKE_CXX_COMPILER=clang++-18 -DCOMPILER_CACHE=sccache -DENABLE_CLANG_TIDY=1 -DENABLE_TESTS=1 -DENABLE_EXAMPLES=1 -DENABLE_UTILS=1 -DENABLE_BUILD_PROFILING=1" -e BUILD_TARGET='all' --volume=/home/ubuntu/.cargo/registry:/rust/cargo/registry  clickhouse/binary-builder:9acb1f664695' returned non-zero exit status 1.
```
I'll just rerun

**Comment by Felixoid** [comment]
@maxknv please, review my changes https://github.com/ClickHouse/ClickHouse/pull/61473/files/c75118c78d52c4e4cf27c213320d049eba049177..758acb433d577e548b2125ee4051996666766df7

**Comment by Felixoid** [comment]
yes, my fixes work https://s3.amazonaws.com/clickhouse-test-reports/61473/758acb433d577e548b2125ee4051996666766df7/bugfix_validation.html

I'll move them to a dedicated PR

File: src/Common/ConcurrencyControl.cpp [hunk 1]
```diff
@@ -103,18 +122,25 @@ ConcurrencyControl::~ConcurrencyControl()
     // Acquire as many slots as we can, but not lower than `min`
     SlotCount granted = std::max(min, std::min(max, available(lock)));
     cur_concurrency += granted;
+    ProfileEvents::increment(ProfileEvents::ConcurrencyControlGrantedHard, min);
```
Comment (KochetovNicolai) [inline src/Common/ConcurrencyControl.cpp:125]: `ConcurrencyControlSlotsGranted`

File: src/Common/ConcurrencyControl.cpp [hunk 2]
```diff
@@ -103,18 +122,25 @@ ConcurrencyControl::~ConcurrencyControl()
...
 
     // Create allocation and start waiting if more slots are required
     if (granted < max)
+    {
+        ProfileEvents::increment(ProfileEvents::ConcurrencyControlGrantDelayed, max - granted);
```
Comment (KochetovNicolai) [inline src/Common/ConcurrencyControl.cpp:130]: `ConcurrencyControlSlotsDelayed`

File: src/Common/ConcurrencyControl.cpp [hunk 3]
```diff
@@ -103,18 +122,25 @@ ConcurrencyControl::~ConcurrencyControl()
...
+        ProfileEvents::increment(ProfileEvents::ConcurrencyControlAllocationDelayed);
```
Comment (KochetovNicolai) [inline src/Common/ConcurrencyControl.cpp:131]: `ConcurrencyControlQueriesDelayed`

**Review by KochetovNicolai — APPROVED** [review]

**Comment by serxa** [comment]
`00814_replicated_minimalistic_part_header_zookeeper` - recently became flaky https://github.com/ClickHouse/ClickHouse/issues/70339

**Comment by serxa** [comment]
`01287_max_execution_speed` - flaky https://github.com/ClickHouse/ClickHouse/issues/69624

**Comment by serxa** [comment]
> 01287_max_execution_speed - flaky https://github.com/ClickHouse/ClickHouse/issues/69624

It blocks merging - I've reset mergeable check manually

**Pull request merged**